### ISSUE TITLE
Moving optimum models and complex demography

### ISSUE THREAD

**Issue by molpopgen** [molpopgen/fwdpy11#1217]
It is possible to inadvertently set up a moving optimum model where the first optimum starts at a time after a deme first exists.  The back end will calculate a non-finite mean fitness for this deme and raise an error from the C++ back end.  While raising the error is good, the message ("non-finite fitnesses encountere") is useless for diagnosing the problem.

Ideally, "genetic value to fitness" maps would be able to be queried with details about deme presence/absence early (in Python) before we hit the C++ back end, raising a useful exception message.

A back and forth elsewhere with @alxsimon led to discovering this issue.

**Comment by molpopgen** [comment]
Idea:

Early in a generation, when we know the sizes of each offspring deme, we should be able to query each genetic value object "does the deme exist"?  There is some subtlety -- models where demes come and go create the possibility of edge cases.

**Comment by molpopgen** [comment]
I suspect that this is best solved by some combo of:

1. Having a general forward -> backward time calculator
2. Using that calculation, plus the current population time point, to...
3. ...scan the demographic model from that time point until the end of the simulation...
4. ...and do "something intelligent" with the genetic value object.

**Comment by molpopgen** [comment]
The ForwardDemesGraph type has `to_forwards_time` and `to_backwards_time` already.

**Comment by molpopgen** [comment]
> We should, I think, be able to extract it as a string from the underlying C++ type and then convert it into a Python `demes.Graph`.

We don't have this functionality yet on the rust side.  An alternative is to take the backwards time and convert it using the time units, etc., in the input graph? I think this needs another issue.

**Comment by molpopgen** [comment]
@alxsimon -- I'm working a PR that defines a new exception class and can generate messages like:

```
deme 2, event fwdpy11.Optimum(optimum=1.0, VS=1.0, when=110) is at time 990 in the past , but the deme starts at 1000
```

**Comment by alxsimon** [comment]
Thanks! And the details of error are very useful to find the culprit.

**Comment by molpopgen** [comment]
I expect this to merge into the `dev` branch within the hour.  I have also updated the error message to report the deme name rather than its integer index.

**Comment by molpopgen** [comment]
Closed via #1299 which has been merged into `dev`.
